src/gt4py/gtc/gtcpp/gtcpp.py

```diff
@@ -10,7 +10,7 @@
 
 
 class Expr(common.Expr):
-    dtype: common.DataType
+    dtype: Optional[common.DataType]
```

Comment:
Because the dtype propagator runs post field validation, we cannot make it non-Optional. Therefore I added a root validator in the root nodes to check that all dtypes are set.